builder/remotecontext/lazycontext.go

```diff
@@ -0,0 +1,116 @@
+package remotecontext
+
+import (
+	"encoding/hex"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/docker/docker/builder"
+	"github.com/docker/docker/pkg/pools"
+	"github.com/docker/docker/pkg/symlink"
+	"github.com/pkg/errors"
+)
+
+// NewLazyContext creates a new LazyContext. LazyContext defines a hashed build
+// context based on a root directory. Individual files are hashed first time
+// they are asked.
+func NewLazyContext(root string) (builder.Remote, error) {
+	return &lazyContext{
+		root: root,
+		sums: make(map[string]string),
+	}, nil
+}
+
+type lazyContext struct {
+	root string
+	sums map[string]string
+}
+
+func (c *lazyContext) Root() string {
+	return c.root
+}
+
+func (c *lazyContext) Close() error {
+	return nil
+}
+
+func (c *lazyContext) Hash(path string) (string, error) {
+	cleanPath, fullPath, err := c.normalize(path)
+	if err != nil {
+		return "", err
+	}
+
+	fi, err := os.Lstat(fullPath)
+	if err != nil {
+		return "", err
+	}
+
+	relPath, err := rel(c.root, fullPath)
+	if err != nil {
+		return "", convertPathError(err, cleanPath)
+	}
+
+	sum, ok := c.sums[relPath]
+	if !ok {
+		sum, err = c.prepareHash(relPath, fi)
+		if err != nil {
+			return "", err
+		}
+	}
+	return sum, nil
+}
+
+func (c *lazyContext) prepareHash(relPath string, fi os.FileInfo) (string, error) {
+	p := filepath.Join(c.root, relPath)
+	h, err := NewFileHash(p, relPath, fi)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create hash for %s", relPath)
+	}
+	if fi.Mode()&(os.ModeDir|os.ModeSymlink|os.ModeDevice|os.ModeSocket|os.ModeNamedPipe) == 0 && fi.Size() > 0 {
+		f, err := os.Open(p)
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to open %s", relPath)
+		}
+		defer f.Close()
+		if _, err := pools.Copy(h, f); err != nil {
+			return "", errors.Wrapf(err, "failed to copy file data for %s", relPath)
+		}
+	}
+	sum := hex.EncodeToString(h.Sum(nil))
+	c.sums[relPath] = sum
+	return sum, nil
+}
+
+func (c *lazyContext) normalize(path string) (cleanPath, fullPath string, err error) {
```

Comment:
Just a reminder from https://github.com/docker/docker/pull/31257#discussion_r107781807 

Also I checked, once you refactored normalize, it is fine to remove the lstat. Any action that was using normalize was doing an Open, or an Lstat, or Remove after it anyways.